acr: support "--repositories" on token creation command

yugangw-msft opened the pull request:
This is to support fast ramp up on repository level of access control. One command will set up all for users. 
Notes:
1. For consistency, I also renamed the `--add` to `--repositories` to be consistent with `az acr token create`, but do let me know if you feel the `--add` is better. 
2. When the existing scope-map is detected, if the actions are the same, CLI will accept it; otherwise it will throw. This is to support command level idempotency that you can run the same command again and again.


- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

hkuadithya commented on src/azure-cli/azure/cli/command_modules/acr/token.py:
```diff
@@ -39,6 +50,53 @@ def acr_token_create(cmd,
+    import re
+    from ._client_factory import cf_acr_scope_maps
+    scope_map_name = DEF_SCOPE_MAP_NAME_TEMPLATE.format(token_name)
+    scope_map_name = re.sub(r'[^a-zA-Z0-9-]', '-', scope_map_name)
```
`scope_map_name` is derived from `<tokenName>-scope-map`.
If there are non permissible characters found in the `<tokenName>`, the token creation itself will fail.

Therefore, it might lead to a case where a scope map gets created but the token creation fails.

@ankurkhemani any thoughts?

yugangw-msft replied:
Do token and scope-name share the same character set?

hkuadithya replied:
Both tokens and scope-,map are ARM resources. Hence, the character set constraints for Scope map and Token is same as the constraints for any other ARM resource.

hkuadithya replied:
Python SDK already has a validation for Token and Scope map resources (both are ARM resources and thus, there's also and existing ARM resource validation in RP).
In my opinion, we shouldn't perform any client side modifications such as    `scope_map_name = re.sub(r'[^a-zA-Z0-9-]', '-', scope_map_name)`.

If creation of `<tokenName>-scope-map` fails, then the operation should be stopped at that point with a suitable error message. 

@ankurkhemani please suggest changes here if needed.

yugangw-msft replied:
Agreed. I will pull out of the client side modification.

yugangw-msft commented on src/azure-cli/azure/cli/command_modules/acr/_params.py:
```diff
@@ -253,21 +253,32 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('scope_map_name', options_list=['--scope-map'], help='The name of the scope map associated with the token', required=False)
-        c.argument('status', options_list=['--status'], help='The status of the token. Allowed values are "enabled" or "disabled".', required=False, default="enabled")
+        c.argument('status', options_list=['--status'], arg_type=get_enum_type(['enabled', 'disabled']),
+                   help='The status of the token', required=False, default="enabled")
```
I am using the enum mainly to leverage CLI's built-in support to show the possible values to user.
Please let me know if I misunderstood.

hkuadithya replied:
I see. Thanks for the clarification.

ankurkhemani replied:
@yugangw-msft You can use the enum from the model and the cli will still show the possible values.
`arg_type=get_enum_type(TokenStatus)`

yugangw-msft commented on src/azure-cli/azure/cli/command_modules/acr/_params.py:
```diff
@@ -253,21 +253,32 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
+        c.argument('scope_map_name', options_list=['--scope-map'],
+                   help='The name of the scope map with pre-configured repository permissions. Use "--repositories" if you would like CLI to configure one for you')
+        c.argument('repositories', nargs='+', action='append',
+                   help='repository permissions. Use the format "--repositories REPO [ACTION1 ACTION2 ...]" per flag. ' + valid_actions)
```
I am changing it to be `--repository` like you recommended which will be consistent with others

hkuadithya replied:
@yugangw-msft 
My comment was about the argument variable `c.argument('repositories'` not the command line argument.  I suggested changing it to a more descriptive variable such as `c.argument('repository_actions_list'` since it's a list of Repositories and their corresponding actions.

I should've been clearer in the review. It's a nit suggestion. Feel free to change it to a more descriptive variable if possible.

yugangw-msft commented on src/azure-cli/azure/cli/command_modules/acr/token.py:
```diff
@@ -2,34 +2,45 @@
+    from ._utils import get_resource_id_by_registry_name
 
-    validate_premium_registry(cmd, registry_name, resource_group_name)
+    if bool(repositories) == bool(scope_map_name):
```
it is possible that both `repositories` and `scope_map_name` could be none, which the if statement also catches. The error of `usage error` is a conventional format, hence let us keep as it is.

hkuadithya replied:
I see. It's like an XOR operation.  At least, repositories  or scope_map_name should be provided, but not both.  Looks good. Thanks for the clarification.

hkuadithya replied:
I see. Thanks for clarifying.

yugangw-msft commented on src/azure-cli/azure/cli/command_modules/acr/_params.py:
```diff
@@ -253,21 +253,32 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
+
     with self.argument_context('acr scope-map create') as c:
-        c.argument('add_repository', options_list=['--add'], nargs='+', help='Actions to be added. Use the format "--add REPO [ACTION1 ACTION2 ...]" per flag. Valid actions are metadata/read, metadata/write, content/read, content/write and content/delete', action='append', required=True)
+        c.argument('repositories', options_list=['--repository'], nargs='+', action='append', required=True,
```
Agreed. I will make the change.

yonzhan commented:
@qwordy please help with this PR.

yugangw-msft commented:
@hkuadithya, @ankurkhemani, I have addressed all feedback, to give doc team enough time to update doc, I am going ahead to merge it, feel free to leave comment if you see any other places I should improve

yugangw-msft commented on src/azure-cli/azure/cli/command_modules/acr/tests/latest/test_acr_token.py:
```diff
@@ -0,0 +1,47 @@
+
+    @AllowLargeResponse()
+    @ResourceGroupPreparer()
+    @unittest.skip('blocked till the feature gets deployed to production')
```
@hkuadithya, this is the test we can re-record &enable once the deployment to prod is done

ankurkhemani commented on src/azure-cli/azure/cli/command_modules/acr/_params.py:
```diff
@@ -254,21 +254,32 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
 
+    valid_actions = "Valid actions are metadata/read, metadata/write, content/read, content/write and content/delete"
+    with self.argument_context('acr scope-map update') as c:
+        c.argument('add_repository', options_list=['--add'], nargs='+', action='append', required=False,
```
So we are still using `--add` `--remove` in update command, but `--repository` in create?

hkuadithya replied:
I see your point. I haven't changed this for now in the new PR.

ankurkhemani commented on src/azure-cli/azure/cli/command_modules/acr/_params.py:
```diff
@@ -254,21 +254,32 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
-        c.argument('remove_repository', options_list=['--remove'], nargs='+', help='Actions to be removed. Use the format "--remove REPO [ACTION1 ACTION2 ...]" per flag. Valid actions are metadata/read, metadata/write, content/read, content/write and content/delete', action='append', required=False)
         c.argument('scope_map_name', options_list=['--name', '-n'], help='The name of the scope map.', required=True)
 
+    valid_actions = "Valid actions are metadata/read, metadata/write, content/read, content/write and content/delete"
```
Would prefer using a defined constant or enum for the actions.

hkuadithya replied:
Done.

ankurkhemani commented on src/azure-cli/azure/cli/command_modules/acr/token.py:
```diff
@@ -39,6 +51,51 @@ def acr_token_create(cmd,
+                           registry_name, scope_map_name))
+    except CloudError:
+        pass
+    logger.warning('Creating a scope map of "%s" for provided respository permissions', scope_map_name)
```
typo: `repository`

ankurkhemani commented on src/azure-cli/azure/cli/command_modules/acr/token.py:
```diff
@@ -39,6 +51,51 @@ def acr_token_create(cmd,
+        pass
+    logger.warning('Creating a scope map of "%s" for provided respository permissions', scope_map_name)
+    poller = scope_map_client.create(resource_group_name, registry_name, scope_map_name,
+                                     actions, "Token {}'s scope map".format(token_name))
```
Not sure if the scope-map `description` is generic enough since one scope-map can be tied to multiple tokens.
How about: `"Created by token: {}".format(token_name)`

ankurkhemani commented:
@yugangw-msft Added some minor comments. LGTM mostly. 

Thanks for this feedback and changes! 
It specially makes sense to also provide one command for token creation and do the magic underneath.
